### PR TITLE
React App - Add html signature download feature

### PR DESCRIPTION
**Pull request by adamnitti**
## overall purpose
adds the ability to download the html of the generated signature

## changes made
- added download button to signature card display
- added `hasRequiredFields` constant
- now copy and download buttons are disabled unless all required fields exist

## screenshot
Screen Shot 2022-12-29 at 3 07 02 PM
Screen Shot 2022-12-29 at 3 07 45 PM

https://user-images.githubusercontent.com/75748078/210011504-4c05b949-fef0-4ec1-a1da-018da53fb4ca.mp4

### CODE REVIEW

File: signature-react-app/src/App.tsx [hunk 1]
```diff
@@ -103,6 +104,13 @@ function App() {
   const classes = useStyles();
   const [state, setState] = React.useState<State>(initialState);
 
+  const hasRequiredFields: boolean =
```
Comment (adamnitti) [inline signature-react-app/src/App.tsx:107]: constant for determining if all required fields are present

File: signature-react-app/src/App.tsx [hunk 2]
```diff
@@ -144,10 +152,14 @@ function App() {
           <br />
           <Button
             onClick={copyToClipboard}
+            disabled={!hasRequiredFields}
             endIcon={state.copied ? <CheckOutlined /> : <FileCopyOutlined />}
           >
             {state.copied ? "Copied" : "Copy to clipboard"}
           </Button>
+          <Button onClick={downloadHtmlFile} disabled={!hasRequiredFields}>
```
Comment (adamnitti) [inline signature-react-app/src/App.tsx]: new button for html file download of signature

File: signature-react-app/src/App.tsx [hunk 3]
```diff
@@ -185,6 +197,28 @@ function App() {
     }
   };
 
+  const downloadHtmlFile = () => {
+    const htmlSignature = ReactDOMServer.renderToStaticMarkup(
```
Comment (adamnitti) [inline signature-react-app/src/App.tsx:205]: grabbing markup for file

File: signature-react-app/src/App.tsx [hunk 4]
```diff
@@ -144,10 +152,14 @@ function App() {
           <br />
           <Button
             onClick={copyToClipboard}
+            disabled={!hasRequiredFields}
```
Comment (adamnitti) [inline signature-react-app/src/App.tsx:155]: disabled button condition for required fields

File: signature-react-app/src/App.tsx (hunk 2 above)
Comment (adamnitti) [inline signature-react-app/src/App.tsx]: disabled button condition for required fields

File: signature-react-app/src/App.tsx [hunk 5]
```diff
@@ -185,6 +197,28 @@ function App() {
...
+        phone={state.phone}
+        mobile={state.mobile}
+        calendlyLink={state.calendlyLink}
+      />
+    );
+    const blob = new Blob([htmlSignature]);
+    const fileDownloadUrl = URL.createObjectURL(blob);
+    const link = document.createElement("a");
+    link.href = fileDownloadUrl;
+    link.setAttribute("download", "signature.html");
```
Comment (chancesmith) [inline signature-react-app/src/App.tsx]: Add person's full name for the file name and `.htm` extension to work with outlook

File: signature-react-app/src/App.tsx [hunk 6]
```diff
@@ -144,10 +152,14 @@ function App() {
...
           <Button
             onClick={copyToClipboard}
+            disabled={!hasRequiredFields}
             endIcon={state.copied ? <CheckOutlined /> : <FileCopyOutlined />}
           >
             {state.copied ? "Copied" : "Copy to clipboard"}
           </Button>
+          <Button onClick={downloadHtmlFile} disabled={!hasRequiredFields}>
+            Download HTML File
+          </Button>
```
Comment (chancesmith) [inline signature-react-app/src/App.tsx:166]: can we add mui download icon?

**Review by chancesmith — CHANGES_REQUESTED** [review]
Looks good. Just a few tweaks.

File: signature-react-app/src/App.tsx (hunk 5 above)
Comment (adamnitti) [reply signature-react-app/src/App.tsx]: done:
Screen Shot 2022-12-29 at 5 11 33 PM

File: signature-react-app/src/App.tsx (hunk 6 above)
Comment (adamnitti) [reply signature-react-app/src/App.tsx:166]: done: 

Screen Shot 2022-12-29 at 5 11 16 PM

File: signature-react-app/src/App.tsx [hunk 7]
```diff
@@ -62,6 +64,9 @@ const useStyles = makeStyles((theme: Theme) =>
       height: 50,
       marginLeft: 0.7,
     },
+    downloadButton: {
+      marginLeft: 5,
+    },
```
Comment (chancesmith) [inline signature-react-app/src/App.tsx]: is this style needed if we use the endIcon prop?

File: signature-react-app/src/App.tsx [hunk 8]
```diff
@@ -144,10 +156,19 @@ function App() {
...
           >
             {state.copied ? "Copied" : "Copy to clipboard"}
           </Button>
+          <Button
+            className={classes.downloadButton}
+            onClick={downloadHtmlFile}
+            disabled={!hasRequiredFields}
+          >
+            Download HTML File
+            <DownloadIcon className={classes.downloadButton} />
```
Comment (chancesmith) [inline signature-react-app/src/App.tsx]: Any reason why we didn't use the endIcon prop on Button for the icon?

File: signature-react-app/src/App.tsx [hunk 9]
```diff
@@ -185,6 +206,31 @@ function App() {
...
+        calendlyLink={state.calendlyLink}
+      />
+    );
+    const blob = new Blob([htmlSignature]);
+    const fileDownloadUrl = URL.createObjectURL(blob);
+    const link = document.createElement("a");
+    link.href = fileDownloadUrl;
+    link.setAttribute(
+      "download",
+      `${state.fullName.split(" ").join("-")}-signature.htm`
```
Comment (chancesmith) [inline signature-react-app/src/App.tsx]: We need to match the output (file naming) of the bulk signature generator, right?

Chad may be scripting the file placement by staff name. We can at least be consistent.

**Review by chancesmith — CHANGES_REQUESTED** [review]

File: signature-react-app/src/App.tsx (hunk 7 above)
Comment (adamnitti) [reply signature-react-app/src/App.tsx]: nah. took it out.

File: signature-react-app/src/App.tsx (hunk 8 above)
Comment (adamnitti) [reply signature-react-app/src/App.tsx]: nope, just put it in there. i hadn't used that prop before.

File: signature-react-app/src/App.tsx (hunk 9 above)
Comment (adamnitti) [reply signature-react-app/src/App.tsx]: Screen Shot 2022-12-30 at 8 50 12 AM

**Review by chancesmith — APPROVED** [review]
🏎️

**Pull request merged**